### ISSUE TITLE
Adding iso 3166-2 support for states

### ISSUE THREAD

**Issue by bvey** [thelia/thelia#3168]
### Description

Thelia state datatable contains only two-character iso codes (ISO 3166 standard).

Adding the five-character iso code support (ISO 3166-2).

For the US states : https://en.wikipedia.org/wiki/ISO_3166-2:US

**Comment by anoziere** [comment]
Triage notes. The full `US-CA` form is already derivable today (`country.isoalpha2` + `state.isocode`), so the actionable gap is convenience, plus the much larger question of worldwide coverage (the seed only ships subdivisions for 7 countries).

Three options weighed:

**A. Computed accessor (shipping now).** `State::getIsoCode3166_2()` returning `{country alpha-2}-{state isocode}`, exposed as a new `ISOCODE_3166_2` output in the state loop and as a read-only field on the API resource. No schema change, no migration.

**B. Repurpose `state.isocode` to store the full code.** Rejected: breaking for every consumer of the bare suffix (the `/admin/states/iso/{countryIso3}/{stateIso}` API route, the address validation traits, 312 seeded rows) for a purely cosmetic gain.

**C. Full worldwide ISO 3166-2 seed (~5,000 rows).** That part is real but belongs in an opt-in geo-data module rather than a 150x growth of the core seed — a module can also track ISO 3166-2 changes independently of Thelia releases. Kept open here as the roadmap discussion.

A is on its way for both branches.

**Comment by anoziere** [comment]
Option A is merged on both lines: #3530 (`2.6`) and #3531 (`main`). `State::getIsoCode3166_2()` returns the full `US-CA` form (null-safe), exposed as `ISOCODE_3166_2` in the state loop and as a read-only `isoCode3166_2` field on the API resource. This issue stays open for the remaining question: worldwide ISO 3166-2 seed coverage (~5,000 rows), which we recommend shipping as a dedicated opt-in geo-data module rather than in the core seed.

**Comment by anoziere** [comment]
## What the seed actually ships today

The remaining half of this issue is a data question, and the starting figure was never written down anywhere. Here it is, counted from `setup/insert.sql` on both lines and compared against the ISO 3166-2 register (`iso-codes` 4.15.0, 2024-01, 5046 entries).

| | `main` (Thelia 3) | `2.6` |
|---|---|---|
| countries seeded | 249 | 249 |
| countries carrying subdivisions | **8** | **7** |
| subdivision rows | **413** | **312** |
| `state_i18n` rows | 2891 (413 × 7 locales) | 2184 |

Per country:

| country | seeded | ISO 3166-2 | matching | missing | seeded codes absent from ISO |
|---|---|---|---|---|---|
| AR Argentina | 24 | 24 | 24 | 0 | — |
| CA Canada | 13 | 13 | 13 | 0 | — |
| JP Japan | 47 | 47 | 47 | 0 | — |
| MX Mexico | 32 | 32 | 30 | 2 | `AGS`, `DIF` |
| US United States | 53 | 57 | 53 | 4 | — (missing AS, GU, MP, UM) |
| ID Indonesia | 33 | 45 | 33 | 12 | — |
| IT Italy | 110 | 126 | 105 | 21 | `AO`, `CI`, `OG`, `OT`, `VS` |
| FR France (`main` only) | 101 | 124 | 100 | 24 | `75` |

**A shop selling to Spain, Germany, the UK, Brazil, Australia, India, the Netherlands, Switzerland or Portugal gets no subdivision at all.** Coverage is 8 countries out of the 200 the standard covers, and 413 subdivisions out of 5046 — 8.2 %.

## Seven codes make `getIsoCode3166_2()` return something that does not exist

Now that `State::getIsoCode3166_2()` (#3530 / #3531) concatenates the country alpha-2 code with `state.isocode`, any wrong suffix becomes a wrong ISO 3166-2 code:

- **`MX-AGS`** — Aguascalientes is `MX-AGU`.
- **`MX-DIF`** — the Federal District became Ciudad de México in 2016 and is `MX-CMX`.
- **`IT-CI`, `IT-OG`, `IT-OT`, `IT-VS`** — Carbonia-Iglesias, Ogliastra, Olbia-Tempio and Medio Campidano were dissolved in 2016 and removed from the register.
- **`IT-AO`** — Aosta is not a separate ISO subdivision; Valle d'Aosta is the autonomous region `IT-23`.
- **`FR-75`** — ISO codes Paris `FR-75C`, the collectivity with special status.

The two Mexican ones are pure renames of the same entity, so they are fixed: #3660 (`main`) and #3661 (`2.6`), seed plus a guarded migration for installed shops, plus a regression test on `main`.

The other five are **not** mechanical and are left for a decision here. The four Sardinian provinces were merged into Sud Sardegna, not renamed, so deleting or remapping the rows would break addresses that point at them. `FR-75` follows the INSEE department numbering #3536 deliberately adopted for the whole French set; moving it alone to `75C` would make France the only country whose codes mix two schemes. My suggestion is to leave all five as they are and treat them as the first concrete evidence that this data needs an owner and a refresh cycle, which is the real subject below.

## What "worldwide coverage" actually costs

The `state` table has **one level**. ISO 3166-2 does not:

- 5046 entries across 200 countries, of which **3590 are top-level** and **1456 are nested** under another subdivision.
- **102 of the 200 countries mix several subdivision types** at the same level.
- The extreme case is the UK: 221 entries, of which 4 are top-level (England, Scotland, Wales, Northern Ireland) and 217 are councils, boroughs and districts. Importing them flat gives a checkout where the buyer picks between "England" and "Barnsley" in the same dropdown.

Thelia's existing seed already made a per-country editorial choice rather than a mechanical one, and it is not the same choice everywhere: the US, Canada, Japan, Mexico and Argentina use the top level, while Italy uses the 110 provinces and France the 101 departments — the second level, because that is the level a postal address is written at. Indonesia likewise seeds the 33 provinces, not the 7 geographical units ISO puts at the top.

So "ship the 5000 subdivisions" is not an import job. It is **200 editorial decisions** about which level a shipping address needs, each of which then has to be kept in step with ISO changes — the 2016 Italian and Mexican reforms above are exactly the kind of drift that has already accumulated.

Order of magnitude for the seed itself: `setup/insert.sql` is 666 KB today. The 3590 top-level subdivisions add roughly 130 KB of `state` rows and, at the 7 locales the file currently carries, roughly 1.1 MB of `state_i18n` rows — a file close to 1.9 MB, and 25 000 `state_i18n` rows inserted on every single install, for shops that overwhelmingly sell in one or two countries.

## Where the data lives, and what that implies

Worth stating because it shapes the answer:

- The seed is `setup/insert.sql` (generated from `setup/insert.sql.tpl`), at the repository root, versioned. `core/bootstrap.php:85-97` makes a root `setup/` win over the `thelia/setup` package, so a change to it takes effect immediately in this repository and its CI. **Unlike `local/config/schema.xml`, the seed is not overwritten by a Composer package during CI** — seed-only pull requests (#3536, #3557, #3567) went green on `main` without any release. So enlarging the seed is technically unblocked today.
- But `setup/` is dispatched to the `thelia/setup` package, and a Composer-installed project reads the package, not this repository. So the data only reaches real shops **on a release** — the core release cadence becomes the ISO update cadence.
- Existing shops need a migration script; `setup/update/sql/*.sql` is the mechanism, and #3536 established the idempotent pattern (`INSERT ... LEFT JOIN existing WHERE existing.id IS NULL`, keyed on `(country_id, isocode)`).

## The two routes

**A. Enlarge the core seed.**
Every install grows by ~1.2 MB of SQL and ~29 000 rows whether or not the shop ships abroad. Every shop already installed needs a migration script that inserts several thousand rows. And the ISO register changes several times a year: each change means a core patch release, and a shop that does not upgrade the core keeps stale subdivisions. Nobody is named as owner of that refresh today, which is precisely how `IT-OG` and `MX-DIF` survived ten years. In exchange: it works out of the box, no discovery problem, and country/state stay in one consistent dataset.

**B. An opt-in geo-data module.**
Nothing is imposed on anyone; a shop selling only in France pays nothing. The module owns the register and can be released whenever ISO publishes a change, independently of the core release train. It can ship per-country activation ("import Spain and Germany"), which is what an integrator actually wants, and it can carry the editorial level choice per country as data rather than as a fait accompli in the core seed. Costs: an integrator has to find out it exists — so it needs to be named in the address/country documentation and ideally surfaced from the back-office country screen; and someone has to decide where it lives and who publishes it.

**Recommendation: B.** The deciding argument is not size, it is ownership. This dataset changes on ISO's schedule, not Thelia's, and the core seed is the one place in the product that cannot be updated without a release. Route A guarantees that the data is stale for most shops most of the time, which is the state we are already in. The 8 countries currently in the seed should stay where they are — removing them would be a breaking change for existing shops — and the module should extend them rather than replace them.

## If B is agreed, what it looks like

Working name `thelia-modules/GeoData` (or `IsoSubdivisions`).

1. **Data source.** Ship the register as a versioned data file in the module, not as SQL — generated from `iso-codes` / the ISO register, with the source version recorded. One entry per subdivision: country alpha-2, ISO suffix, parent, type, English name.
2. **Per-country editorial level.** A small curated map deciding, for each country, which level is the postal level (top level by default; second level for IT, FR, ID, GB, ES, PH and the other nested cases). This is the only part that needs judgement, and it can be filled in progressively.
3. **Import command.** `module:geo-data:import --country=ES,DE` — idempotent, keyed on `(country_id, isocode)`, never touching a row that already exists, and setting `country.has_states` only when the integrator asks for it. Same guarantees as the #3536 migration.
4. **No schema change.** `state.isocode` is `VARCHAR(4)` (`local/config/schema.xml`), and the longest ISO 3166-2 suffix in the whole register is 3 characters (327 of length 1, 3099 of length 2, 1620 of length 3), so everything fits as is. Names go to `state_i18n`; locales beyond English fall back to the English name rather than being invented.
5. **Refresh.** A second command reporting the diff between the shipped register and the shop's rows (added, renamed, withdrawn) without applying anything, so an exploitant decides. Withdrawn subdivisions are reported, never deleted — addresses point at them.
6. **Discovery.** A line in the address/country documentation and a mention on the back-office country page.

Two things that would still belong to the core, and are out of the module's scope: `country.has_states` currently gates the whole state feature (`Thelia\Form\AddressCountryValidationTrait:51`, `Thelia\Tools\AddressFormat:152`), and it is a boolean — there is no way to say "a state is optional here", which is why France ships 101 departments with `has_states = 0` and no form offers them. And the seed's `state_i18n` only carries 7 of the 18 locales under `setup/I18n/`, with de_DE, es_ES, cs_CZ and it_IT filled for a quarter of the rows, because `generate:sql` is broken (noted in #3536). Both are worth their own issues.

## Summary

- Census done: 8 countries, 413 subdivisions, 8.2 % of ISO 3166-2; the large European and Asian markets have none.
- 7 seeded codes are not valid ISO 3166-2; the 2 that are pure renames are fixed in #3660 / #3661, the other 5 need a decision.
- Recommendation: an opt-in geo-data module, because the register changes on ISO's schedule and the core seed can only change on a release.

**Comment by anoziere** [comment]
Two pull requests merged, https://github.com/thelia/thelia/pull/3660 (`main`) and https://github.com/thelia/thelia/pull/3661 (`2.6`), and the census this issue was missing.

**The census.** Counted from the seed against the ISO 3166-2 register (5046 entries): **8 countries out of 200 carry subdivisions, 413 rows out of 5046 — 8.2 %**. Argentina, Canada and Japan are complete; the United States has 53 of 57; Italy 110 of 126; Indonesia 33 of 45. Spain, Germany, the United Kingdom, Brazil, Australia, India, the Netherlands, Switzerland and Portugal have **none**. So an integrator selling to the US, Canada or Italy is served, and one selling to Spain gets nothing.

**Seven seeded codes are not valid ISO 3166-2**, which matters now that `State::getIsoCode3166_2()` concatenates the country code with them: `MX-AGS` (should be `AGU`), `MX-DIF` (Ciudad de México since 2016, `CMX`), `IT-CI`, `IT-OG`, `IT-OT`, `IT-VS` (Sardinian provinces dissolved in 2016) and `FR-75` (ISO uses `75C`). They come from `setup/update/sql/2.3.0-alpha1.sql`.

*Edited: `IT-AO` was listed here by mistake, which made the list show eight entries under the word seven. ISO 3166-2:IT does define `AO` for the Aosta province, so the code is valid and the seven above are the real ones.*

Only the **two pure renames** were fixed, with an idempotent migration that is a no-op on a shop which already corrected it. The other five are merges of dissolved provinces or a deliberate INSEE codification choice: patching them silently would break stored addresses or make France mix two schemes. **They need a decision**, they are listed in the pull request, and nothing was touched.

**On delivery mechanics, checked rather than assumed**: the seed does *not* behave like `schema.xml`. The root `setup/` wins over the `thelia/setup` package, which extracts to a different, gitignored path, so a seed-only pull request does not break CI — confirmed by #3536, #3557 and #3567 going green with no release. But `setup/` *is* dispatched to the package, so seed data reaches real shops **only on a release**: the core release cadence becomes the ISO update cadence. No schema change was needed either — `state.isocode` is `VARCHAR(4)` and the longest ISO suffix in the whole register is 3 characters, so even full worldwide coverage would need no column.

**Recommendation: an opt-in module**, and the decisive argument is not the volume. ISO 3166-2 is **not a flat list** — 3590 top-level entries, 1456 nested, and 102 of the 200 countries mix several subdivision types at one level. The United Kingdom is 221 entries of which 4 are countries and 217 are councils; imported flat, a checkout would offer "England" and "Barnsley" side by side. Thelia own seed already made a per-country editorial choice, and not the same one everywhere: the US, Canada, Japan, Mexico and Argentina use the top level, while Italy, France and Indonesia use the second, because that is the level a postal address uses. "Ship the 5000 rows" is therefore 200 editorial decisions, each needing upkeep — which is exactly how `IT-OG` and `MX-DIF` survived ten years in the seed.

No repository was created. Two side findings worth their own tickets: `country.has_states` is a boolean with no "optional state" value, which is why France ships 101 departments that no address form offers; and `state_i18n` covers 7 of the 18 locales because `generate:sql` is broken, already noted in #3536.

**Comment by anoziere** [comment]
Closing this one. What it asked for is in, and what it grew into belongs elsewhere.

**Delivered.** `State::getIsoCode3166_2()` returns the five-character code, the `state` loop exposes `$ISOCODE_3166_2` and the API resource exposes `isoCode3166_2` in read: #3531 on `main`, #3530 on `2.6`. No column and no migration were needed — `state.isocode` is already a `VARCHAR(4)` and the longest suffix in the whole register is three characters. The loop output is now documented in the three output tables of thelia/docs (thelia/docs#29), which it was not.

**Seed corrections that came out of the census.** #3660 and #3661 fixed the two Mexican codes (`AGS` to `AGU`, `DIF` to `CMX`). #3690 and #3693 finish the job: Ciudad de México was still titled Distrito Federal in English, French and Russian, and the four Sardinian provinces dissolved in 2016 are now seeded `visible = 0` with `IT-SU` (Sud Sardegna) added, rather than deleted, so addresses that already point at them stay readable. Each correction ships a guarded statement for installed shops, matched on the outdated value only.

Two codes were deliberately left alone. `IT-AO` is valid — ISO 3166-2:IT does define `AO` for Aosta, it was wrongly listed above and the comment is corrected. `FR-75` follows the INSEE numbering chosen for the French departments in #3536; changing it would make France the only country mixing two schemes.

**Worldwide coverage goes to an opt-in community module, not to the core.** The seed covers 413 subdivisions out of 5046, 8.2 %, across 8 countries. Growing that in the core would tie the ISO update cadence to the core release cadence, and that is precisely the mechanism that let `IT-OG` and `MX-DIF` sit in the seed for ten years. It would also push 200 editorial decisions into the core: the register is not a flat list, and Thelia own seed already picks a different level per country because that is the level a postal address uses.

The honest counter-argument is that this route has been taken twice already. [FrenchDepartments](https://github.com/thelia-modules/FrenchDepartments) and [INSEEGeo](https://github.com/thelia-modules/INSEEGeo) both do a one-shot geographic import, and both are dormant, since 2021 and 2016. And #3536 moved the other way three days ago by bringing the French departments into the core seed. France is the exception there because it is the default country of the distribution and its departments are wanted out of the box; that is not an argument for shipping the other 199. A module without a named maintainer would be a third dormant repository, so the module is the plan and an owner is the condition.

**Two side findings are now tickets** rather than paragraphs in this thread: #3696 on `country.has_states` having no "optional state" value, which is why France carries 101 departments that no address form offers, and #3697 on the seed covering 7 of the 18 languages shipped in `setup/I18n`.

**One divergence is deliberate and worth writing down.** `main` seeds 414 states, `2.6` seeds 313. The 101 French departments of #3536 are not backported: they are new data, not a correction, and `2.6` is a stable line. Both branches carry the same corrections, not the same catalogue.
